compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/body/resolve/FirDeclarationsResolveTransformer.kt

```diff
@@ -271,12 +274,17 @@ class FirDeclarationsResolveTransformer(transformer: FirBodyResolveTransformer)
                 lambdaResolution.expectedReturnTypeRef ?: anonymousFunction.returnTypeRef.takeUnless { it is FirImplicitTypeRef }
             val result = transformFunction(anonymousFunction, withExpectedType(expectedReturnType)).single as FirAnonymousFunction
             val body = result.body
-            return if (result.returnTypeRef is FirImplicitTypeRef && body != null) {
+            if (result.returnTypeRef is FirImplicitTypeRef && body != null) {
                 result.transformReturnTypeRef(transformer, withExpectedType(body.resultType))
-                result
-            } else {
-                result
             }
+            val statement = body?.statements?.lastOrNull()
+            if (statement is FirExpression && statement !is FirReturnExpression && !body.resultType.isNothing) {
+                val returnExpression = FirReturnExpressionImpl(statement.source, statement)
+                returnExpression.target = FirFunctionTarget(null)
+                returnExpression.target.bind(result)
+                (body.statements as? MutableList<FirStatement>)?.set(body.statements.size - 1, returnExpression)
```

Comment:
This is very hacky way, moreover it will work incorrectly for anonymous functions with expression bodies. It's better to do it via transformer (`LastBodyStatementTransformer`?) instead.

Reply:
Moved to Fir2Ir.

Also would you please elaborate on "work incorrectly for anonymous functions with expression bodies"? I tried an example `val f = {"K"}`: currently the frontend generates a function body with a single FirConstExpression, and thus inserting a return wrapper would work. Are there other cases you'd expect not to insert return?

Reply:
No, the example should look like `callOfFunctionWithLambda(x, y, fun(n: Int) = 42 * n)`.

Reply:
As discussed today, this should be kept in FIR itself, but utility transformer like `StoreType` should be used.

Reply:
Moved back to Fir, using a transformer